JsonLow.d.ts

```diff
@@ -1,16 +1,91 @@
+export declare const CodePoint: CodePoint
+export type CodePoint = {
+  _0_: number,
+  _1_: number,
+  _9_: number,
+  _a_: number,
+  _f_: number,
+  _A_: number,
+  _F_: number,
+  _openCurly_: number,
+  _openSquare_: number,
+  _closeCurly_: number,
+  _closeSquare_: number,
+  _quoteMark_: number,
+  _plus_: number,
+  _minus_: number,
+  _space_: number,
+  _newline_: number,
+  _tab_: number,
+  _return_: number,
+  _backslash_: number,
+  _slash_: number,
+  _comma_: number,
+  _colon_: number,
+  _t_: number,
+  _n_: number,
+  _b_: number,
+  _r_: number,
+  _u_: number,
+  _dot_: number,
+  _e_: number,
+  _E_: number,
+  _l_: number,
+  _s_: number,
+}
+export enum JsonFeedbackType {
+  error = 'JsonFeedbackType.error',
+}
+export enum JsonErrorType {
+  unexpected = 'JsonErrorType.unexpected',
+  unexpectedEnd = 'JsonErrorType.unexpectedEnd',
+}
+export declare const error: error
+export type error = (message: string) => {
+  type: JsonFeedbackType.error,
+  message: string,
+}
+export declare const unexpected: unexpected
+export type unexpected = (code: number, context: string, expected: Array<string | [startChar: string, endChar: string]>) => {
+  type: JsonFeedbackType.error,
+  errorType: JsonErrorType.unexpected,
+  codePoint: number,
+  context: string,
+  expected: Array<string | [startChar: string, endChar: string]>,
+}
+export declare const unexpectedEnd: unexpectedEnd
+export type unexpectedEnd = (context?: string, expected?: Array<string | [startChar: string, endChar: string]>) => {
+  type: JsonFeedbackType.error,
+  errorType: JsonErrorType.unexpectedEnd,
+  context?: string,
+  expected?: Array<string | [startChar: string, endChar: string]>,
+}
+export declare const isZeroNine: isZeroNine
+export type isZeroNine = (code: number) => boolean
+export declare const isOneNine: isOneNine
+export type isOneNine = (code: number) => boolean
+export declare const isWhitespace: isWhitespace
+export type isWhitespace = (code: number) => boolean
+export interface JsonLowBaseState {
+  mode: string,
+  parents: string[],
+  isKey: boolean,
+  hexIndex: number,
+  maxDepth: number,
+}
+export interface JsonLowState extends JsonLowBaseState {
+  downstream?: JsonLowState,
```

Comment:
(I am semi-thinking out loud here, so bear with me)

That the `downstream` property type matches the `JsonLowHandlers.next`'s return type is correct.

However this type should be either generic or `unknown` (depending on how pragmatic we want to be). The type is actually determined by the handler and so it can't be known statically until we have a specific handler -- in other words we only find out this type when we construct an instance of `JsonLow`, providing it with the `next` parameter.

So if we make this type generic we should add a corresponding generic type parameter to `JsonLowHandlers`, as in `JsonLowHandlers<Feedback, End, T>`.

Or just make it `unknown` and be done with it.

Leaving it up to you, depending on how the level of accuracy you wish to have.

Reply:
Sorry for the late response. I made downstream generic, but I also made it default to `unknown` in case the user doesn't care about the type

Reply:
Another thing: `codePoint` now returns `Feedback | JsonStandardFeedback | undefined`. This is because there are some standard feedback types which are returned, but they are not included as part of the return value, which causes you to have to cast stuff to the `Feedback` type. Another problem is that there might not be a handler, meaning that the return value may be undefined, hence the inclusion of that in the return type.

The `end` method had the same problem, so the same thing was done for that method